lib/puppet/functions/ensure_at_least.rb

```diff
@@ -0,0 +1,72 @@
+# @summary
+#   Return the package version for a given package that is at least the given minimum version
+#   The function will compare the currently installed package with the given minimum version and:
+#    - return the minimum version if the installed package is less than the minimum version, or not present
+#    - return the installed version if the installed package is the same or newer than the minimum version
+#
+# @example Example usage:
+#   # Ensure at least version 2.7.5-77.el7_6 of python is installed
+#     package { 'python':
+#       ensure => ensure_at_least('python', '2.7.5-77.el7_6')
+#     }
+#
+# @example To specify a specific provider:
+#   # Ensure at least version 2.7.5-77.el7_6 of python is installed
+#     package { 'openssl':
+#       ensure   => ensure_at_least('openssl', '2.1.2', 'gem'),
+#       provider => gem
+#     }
+#
+Puppet::Functions.create_function(:ensure_at_least) do
+  # @param package
+  #   The package to ensure
+  #
+  # @param minversion
+  #   The minimum version to ensure
+  #
+  # @param provider
+  #   (Optional) The provider to use
+  #
+  # @return [String] package version value.
+  dispatch :ensure_at_least do
+    required_param 'String', :package
+    required_param 'String', :minversion
+    optional_param 'String', :provider
+    return_type 'String'
+  end
+
+  def ensure_at_least(package, minversion, provider = nil)
+    trusted = closure_scope['trusted']
+    query = [
+      'package_inventory[version] {',
+      "certname = '#{trusted['certname']}'",
+      "and package_name = '#{package}'",
+      ("and provider = '#{provider}'" unless provider.nil?),
+      '}',
+    ].compact.join(' ')
+
+    queryresult = call_function('puppetdb_query', query)
+    if queryresult.length > 1
+      if provider.nil?
+        # Assume first PDB query result is for the default provider
```

Comment:
We should not assume. If there are multiple results returned and no provider specified, we should raise an error.

Better to fail catalog compilation than to possibly mis-configure a system.

Reply:
While I could replicate the logic from the package provider to determine the standard provider when none is specified (based on the OS of the node), it would yield worse results compared to letting PuppetDB handle it. For example, PuppetDB will return packages from the chocolatey provider without issue, even though this is not the default provider on Windows. If we used traditional OS-based logic to determine the default provider when none is specified, the wrong provider would be selected in more cases. Selecting the first record from the PuppetDB results will ensure a valid provider is always chosen.

Multiple results can also come from multiple package versions being installed side-by-side, all from the same provider. That should not cause a compilation error.

Reply:
The PuppetDB team has confirmed that no intentional sorting is happening, it seems to be coming from Postgres' internals at the moment. The suggestion is to leverage `Gem::Version` of that is available. I will have a look at it.

Reply:
The PuppetDB team has confirmed that no intentional sorting is happening, it seems to be coming from Postgres' internals at the moment. No logic exists in PuppetDB on default/primary providers for a resource, this all happens outside of PuppetDB. I will have a look at pulling in the default provider logic from the package resource provider, with some additional logic to fallback to the record returned from PuppetDB in safe scenarios.

Reply:
Default provider logic replicated from the Puppet's package resource providers. Enriched with auto-selecting providers in known-safe scenarios.